consensus/fork_choice/src/fork_choice.rs

```diff
@@ -737,43 +760,80 @@ where
+                | BeaconBlockRef::Merge(_)
+                | BeaconBlockRef::Altair(_) => match progressive_balances_mode {
+                    ProgressiveBalancesMode::Disabled => {
+                        let participation_cache = ParticipationCache::new(state, spec)
+                            .map_err(Error::ParticipationCacheBuild)?;
                         per_epoch_processing::altair::process_justification_and_finalization(
                             state,
                             &participation_cache,
                         )?
                     }
-                    BeaconBlockRef::Base(_) => {
-                        let mut validator_statuses =
-                            per_epoch_processing::base::ValidatorStatuses::new(state, spec)
-                                .map_err(Error::ValidatorStatuses)?;
-                        validator_statuses
-                            .process_attestations(state)
-                            .map_err(Error::ValidatorStatuses)?;
-                        per_epoch_processing::base::process_justification_and_finalization(
-                            state,
-                            &validator_statuses.total_balances,
-                            spec,
-                        )?
+                    ProgressiveBalancesMode::Fast
+                    | ProgressiveBalancesMode::Checked
+                    | ProgressiveBalancesMode::Strict => {
+                        let maybe_participation_cache = progressive_balances_mode
+                            .perform_comparative_checks()
+                            .then(|| {
+                                ParticipationCache::new(state, spec)
+                                    .map_err(Error::ParticipationCacheBuild)
+                            })
+                            .transpose()?;
+
+                        process_justification_and_finalization_from_progressive_cache::<E, T>(
+                                state,
+                                maybe_participation_cache.as_ref(),
+                            )
+                            .or_else(|e| {
+                                if progressive_balances_mode == ProgressiveBalancesMode::Checked {
+                                    error!(
+                                        log,
+                                        "Processing with progressive balances cache failed in checked mode";
+                                        "info" => "falling back to the non-optimized processing method",
+                                        "error" => ?e,
+                                    );
+                                    per_epoch_processing::altair::process_justification_and_finalization(
+                                        state,
+                                        maybe_participation_cache.as_ref().ok_or(Error::ParticipationCacheMissing)?,
+                                    ).map_err(Error::from)
```

Comment:
```suggestion
                                if progressive_balances_mode != ProgressiveBalancesMode::Strict {
                                    error!(
                                        log,
                                        "Processing with progressive balances cache failed in checked mode";
                                        "info" => "falling back to the non-optimized processing method",
                                        "error" => ?e,
                                    );
                                    let participation_cache = maybe_participation_cache
                                        .map(Result::Ok)
                                        .unwrap_or_else(|| ParticipationCache::new(state, spec))
                                        .map_err(Error::ParticipationCacheBuild)?;
                                    per_epoch_processing::altair::process_justification_and_finalization(
                                        state,
                                        &participation_cache,
                                    ).map_err(Error::from)
```

We could be a little bit more paranoid here and *always* fall back to the old method if the new method fails (unless we're in `Strict` mode)?

Reply:
Nice, being paranoid is good for us - I need more of this actually, it's a bit of a mindset shift to what I was used to for a while!
This is definitely much better error handling! (and the `unwrap_or_else`), will update to the suggested approach and remove the unnecessary `ParticipationCacheMissing` error variant.

Reply:
Updated in https://github.com/sigp/lighthouse/pull/4362/commits/90acdc720c4e6db1c5ee9ea4c120cf48489fce4d. Thanks 🙏